### PR TITLE
MoE: remove the replicated-activation repeat in RoutedMoE.permute

### PR DESCRIPTION
**Pull request by systalyze-ai**
# Description

Changes the non-ragged `RoutedMoE.permute` path to gather directly from the original activations instead of materializing `top_k` copies before the sorted gather. This removes an activation buffer that is 6x the dispatch input for DeepSeek-V4-284B.

# Performance

Step time changed by ~1% on v6e-128 with DeepSeek-V4-Flash 284B LoRA fine-tuning at seq 16384; the benefit of the change is the removed activation buffer. Setup: `ici_expert_parallelism=16`, `ici_tensor_parallelism=8`, `ici_fsdp_parallelism=1`, `max_target_length=16384`, `per_device_batch_size=0.125` (global batch 16), `LIBTPU_INIT_ARGS=--xla_tpu_scoped_vmem_limit_kib=98304`, dynamic-splash attention path enabled.

# Tests

- `JAX_PLATFORMS=cpu pytest -q tests/unit/moe_test.py -k GatherReplicatedActivationsTest`: 2 passed.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -171,6 +171,51 @@ def _sort_activations_custom_bwd(residuals: jax.Array, grads: jax.Array) -> tupl
...
+    inputs: jax.Array, sort_indices: jax.Array, num_repeats: int
+) -> tuple[jax.Array, jax.Array]:
+  """Forward pass of the custom vjp for `_gather_replicated_activations()`."""
+  return _gather_replicated_custom(inputs, sort_indices, num_repeats), sort_indices
+
+
+def _gather_replicated_custom_bwd(num_repeats: int, residuals: jax.Array, grads: jax.Array) -> tuple[jax.Array, None]:
+  """Backward pass of the custom vjp for `_gather_replicated_activations()`."""
+  sort_indices = residuals
+  inverse_indices = jnp.empty_like(sort_indices).at[sort_indices].set(jnp.arange(sort_indices.shape[0]))
```
[the remark this answers is not in the document]
Comment (RSudarsanan) [reply src/maxtext/layers/moe.py]: I fixed this, can you check @gemini-code-assist

**Review by richjames0 — CHANGES_REQUESTED** [review]
This seems to have been largely superseded by an equivalent change on main.

`4913 replaces jnp.repeat(inputs_2d, top_k, axis=0) + gather with a fused inputs[sort_indices // num_repeats]. Main already does exactly this, behind a config flag - see (moe.py:1249)`

That said it looks like it would be worth saving as a standalone PR

```
_gather_replicated_custom_bwd computes the inverse permutation with an O(N) scatter:

inverse_indices = jnp.zeros_like(sort_indices).at[sort_indices].set(jnp.arange(sort_indices.shape[0]))

whereas main's _sort_activations_custom_bwd (moe.py:182-185) still pays an O(N log N) jnp.argsort(sort_indices).
```
